homeassistant/components/tado/water_heater.py

```diff
@@ -0,0 +1,303 @@
+            # Fallback to Smart Schedule at next Schedule switch
+            self._default_overlay = CONST_OVERLAY_TADO_MODE
+        else:
+            _LOGGER.debug("Default overlay is set to MANUAL MODE")
+            # Don't fallback to Smart Schedule, but keep in manual mode
+            self._default_overlay = CONST_OVERLAY_MANUAL
+
+        self._current_operation = CONST_MODE_SMART_SCHEDULE
+        self._overlay_mode = CONST_MODE_SMART_SCHEDULE
+
+    async def async_added_to_hass(self):
+        """Register for sensor updates."""
+
+        @callback
+        def async_update_callback():
+            """Schedule an entity update."""
+            self.async_schedule_update_ha_state(True)
+
+        async_dispatcher_connect(
+            self.hass,
+            SIGNAL_TADO_UPDATE_RECEIVED.format("zone", self.zone_id),
+            async_update_callback,
+        )
+
+    @property
+    def supported_features(self):
+        """Return the list of supported features."""
+        return self._supported_features
+
+    @property
+    def name(self):
+        """Return the name of the entity."""
+        return self.zone_name
+
+    @property
+    def unique_id(self):
+        """Return the unique id."""
+        return self._unique_id
+
+    @property
+    def current_operation(self):
+        """Return current readable operation mode."""
+        return WATER_HEATER_MAP_TADO.get(self._current_operation)
+
+    @property
+    def target_temperature(self):
+        """Return the temperature we try to reach."""
+        return self._target_temp
+
+    @property
+    def is_away_mode_on(self):
+        """Return true if away mode is on."""
+        return self._is_away
+
+    @property
+    def operation_list(self):
+        """Return the list of available operation modes (readable)."""
+        return [MODE_AUTO, MODE_HEAT, MODE_OFF]
+
+    @property
+    def temperature_unit(self):
+        """Return the unit of measurement used by the platform."""
+        return TEMP_CELSIUS
+
+    @property
+    def min_temp(self):
+        """Return the minimum temperature."""
+        return self._min_temperature
+
+    @property
+    def max_temp(self):
+        """Return the maximum temperature."""
+        return self._max_temperature
+
+    def set_operation_mode(self, operation_mode):
+        """Set new operation mode."""
+        mode = None
+
+        if operation_mode == MODE_OFF:
+            mode = CONST_MODE_OFF
+        elif operation_mode == MODE_AUTO:
+            mode = CONST_MODE_SMART_SCHEDULE
+        elif operation_mode == MODE_HEAT:
+            mode = self._default_overlay
+
+        self._current_operation = mode
+        self._overlay_mode = None
+
+        # Set a target temperature if we don't have any
+        if mode == CONST_OVERLAY_TADO_MODE and self._target_temp is None:
+            self._target_temp = self.min_temp
+            self.schedule_update_ha_state()
```

Comment:
Do we need an extra state update scheduled? Since this entity is polling, there will by default be a state update scheduled at the end of each service call.

Reply:
The platform itself isn't polling, we use the dispatcher to notify the platforms for new data.

Reply:
When this is left out, the state isn't updated in home assistant, and I only see one request to make the api call.

Reply:
The platform is "polling" since by default the `should_poll` entity property is set to `True`.

https://github.com/home-assistant/home-assistant/blob/9eb04152342609b076837ad8d8fef380f1c46251/homeassistant/helpers/entity.py#L122-L128

https://github.com/home-assistant/home-assistant/blob/9eb04152342609b076837ad8d8fef380f1c46251/homeassistant/helpers/service.py#L406-L419

Reply:
So what happens in `update` will determine the final state since the core calls `entity.async_update_ha_state(True)`.

Reply:
Hmm, it seems I've missed the `should_poll`. It definitely should be `False` (as it is in the `climate` platform.

The update happens in the component trough `hass.helpers.event.track_time_interval`, and that uses the dispatcher to notify the platforms instantly when new data is available. There, a `self.async_schedule_update_ha_state(True)` is called that calls the `update` method.

Reply:
Hmm, okay, there is something else going on here. It seems the state is registered fine without this extra `self.schedule_update_ha_state()` call, but the frontend isn't updated automatically. When I do a `self.schedule_update_ha_state()`, the frontend is automatically updated fine.

Reply:
I've removed this line. It was working fine for `water_heater`.